File: python/tests/test_subscribe_after_start.py
# -*- coding: utf-8 -*-
"""订阅后置 + 立即心跳（#73）离线单测，不需要集群。

对齐基准（Java 5.5.1 ``DefaultMQPushConsumerImpl``）：
  * ``subscribe(topic, subExpression):1265-1275``、class-filter 变体 ``:1277-1287``、
    MessageSelector 变体 —— 三者都是 put 进 ``subscriptionInner`` 之后
    ``if (this.mQClientFactory != null) this.mQClientFactory.sendHeartbeatToAllBrokerWithLock();``
    **没有**「started 之后禁止订阅」这道闸门，且心跳是**同步立即**发的。
  * ``unsubscribe(topic):1317-1319`` 只 remove，**不**发心跳。
  * 心跳体 ``subscriptions():1385-1387`` = ``subscriptionInner.values()``，即**当前**订阅表。

为什么必须离线锁死：真机上「新订阅没推给 broker」是静默的 —— 只表现为
``QUERY_TOPIC_CONSUME_BY_WHO(300)`` 查不到本组、新 topic 分不到队列，客户端一声不响。
这里用替身 client 把**真实**心跳报文体（``_build_heartbeat`` 造的）截下来，
断言里面的订阅集就是刚写进去的那一份，以及 ``unsubscribe`` 一轮都不多发。
"""
from __future__ import annotations

from typing import List

from rocketmq.client.consumer import DefaultMQPushConsumer, MessageSelector
from rocketmq.common.subscription_data import ExpressionType
from rocketmq.remoting.protocol.heartbeat import HeartbeatData

TOPIC = "SubAfterStartTopic"


class FakeClient:
    """只实现心跳路径要用的两个方法：截报文体 + 记录路由登记。"""

    def __init__(self) -> None:
        self.heartbeats: List[HeartbeatData] = []
        self.topics_in_use: List[str] = []

    def get_route_of_all_brokers(self) -> List[str]:
        return ["127.0.0.1:10911"]

    def send_heartbeat(self, addr: str, hb: HeartbeatData, timeout_millis: int) -> None:
        self.heartbeats.append(hb)

    def register_topic_in_use(self, topic: str) -> None:
        self.topics_in_use.append(topic)


def started(group: str = "GID_SubAfterStart"):
    """造一个「已启动」的消费者，client 是替身，心跳报文可截获。"""
    c = DefaultMQPushConsumer(group)
    fake = FakeClient()
    c._mq_client = fake
    c._started = True
    return c, fake


def subs_of(hb: HeartbeatData):
    """心跳体里那一份 ConsumerData 的订阅表（topic → SubscriptionData）。"""
    cd = list(hb.consumer_data_set)[0]
    return {s.topic: s for s in cd.subscription_data_set}


def test_subscribe_after_start_reaches_the_live_table_and_pushes_one_heartbeat():
    """Java :1265-1275：start() 之后 put 完立即发一轮心跳，报文里带上新订阅。"""
    c, fake = started()

    c.subscribe(TOPIC, "TagA||TagB")

    assert len(fake.heartbeats) == 1, "一次 subscribe 恰好一轮心跳"
    body = subs_of(fake.heartbeats[0])
    assert TOPIC in body, "心跳报文必须带新订阅（否则 broker 记不到 topicGroupTable）"
    assert body[TOPIC].sub_string == "TagA||TagB"
    assert body[TOPIC].tags_set == {"TagA", "TagB"}
    # 路由刷新任务要覆盖新 topic（对应 Java 周期任务遍历 live 订阅表收集 topic）
    assert fake.topics_in_use == [TOPIC]
    assert [s.topic for s in c.subscriptions()] == [TOPIC]


def test_selector_subscribe_after_start_keeps_expression_type_in_the_body():
    """Java :1289-1303：MessageSelector 变体同样立即发心跳，且 expressionType 照原样。"""
    c, fake = started()
    c.subscribe_with_selector(TOPIC, MessageSelector.by_sql("a > 1"))

    assert len(fake.heartbeats) == 1
    body = subs_of(fake.heartbeats[0])
    assert body[TOPIC].expression_type == ExpressionType.SQL92
    assert body[TOPIC].sub_string == "a > 1"
    assert not body[TOPIC].tags_set, "SQL92 不带 tagsSet"

    c2, fake2 = started("GID_SubAfterStart2")
    c2.subscribe_with_selector(TOPIC, MessageSelector.by_tag("X||Y"))
    assert len(fake2.heartbeats) == 1
    tag_body = subs_of(fake2.heartbeats[0])[TOPIC]
    assert tag_body.expression_type == ExpressionType.TAG
    assert tag_body.tags_set == {"X", "Y"}


def test_subscribe_before_start_only_records():
    """Java 的 ``mQClientFactory == null`` 分支：没启动就只落订阅表，不碰网络。"""
    c = DefaultMQPushConsumer("GID_SubBeforeStart")
    c.subscribe(TOPIC, "*")
    assert [s.topic for s in c.subscriptions()] == [TOPIC]

    # 就算 client 已经建好、只是还没 start()，也不发（Java 的门是 mQClientFactory != null，
    # 本端口的门是 _started and _mq_client —— start() 末尾才会同时成立）。
    fake = FakeClient()
    c._mq_client = fake
    c._started = False
    c.subscribe("T2", "*")
    assert fake.heartbeats == []
    assert fake.topics_in_use == []


def test_unsubscribe_does_not_push_a_heartbeat():
    """Java :1317-1319：只删表项。多发一轮心跳会掩盖「订阅已撤但 broker 仍记着」的差异。"""
    c, fake = started()
    c.subscribe(TOPIC, "*")
    assert len(fake.heartbeats) == 1

    c.unsubscribe(TOPIC)

    assert len(fake.heartbeats) == 1, "unsubscribe 不发心跳"
    assert c.subscriptions() == []


def test_every_heartbeat_is_rebuilt_from_the_live_table():
    """心跳体是「发出时刻」的订阅表快照，不是启动时的固定集合（Java subscriptions():1385）。"""
    c, fake = started()
    c.subscribe("T_A", "*")
    c.subscribe("T_B", "*")
    c.unsubscribe("T_A")

    c._send_heartbeat_to_all_broker()

    latest = subs_of(fake.heartbeats[-1])
    assert set(latest) == {"T_B"}, "撤掉的 T_A 不能再出现在报文里"
    assert set(subs_of(fake.heartbeats[0])) == {"T_A"}


def test_resubscribe_same_topic_replaces_and_heartbeats_again():
    """Java 每次 subscribe 都是 put + 发心跳（覆盖写也算一次订阅变更）。"""
    c, fake = started()
    c.subscribe(TOPIC, "A")
    c.subscribe(TOPIC, "B")

    assert len(fake.heartbeats) == 2
    assert subs_of(fake.heartbeats[-1])[TOPIC].sub_string == "B"
    cd = list(fake.heartbeats[-1].consumer_data_set)[0]
    assert len(cd.subscription_data_set) == 1, "同一 topic 覆盖，不是两条"

File: cpp/tests/test_subscribe_after_start.cpp
// 订阅后置 + 立即心跳（#73）离线单测 —— 不需要集群。
//
// 对齐基准（Java 5.5.1 `DefaultMQPushConsumerImpl`）：
//   * `subscribe(topic, subExpression):1265-1275`（以及 class-filter / MessageSelector 两个
//     重载）都是 put 进 `subscriptionInner` 之后 `if (mQClientFactory != null)
//     mQClientFactory.sendHeartbeatToAllBrokerWithLock();` —— **没有**「started 之后禁止
//     订阅」这道闸门，心跳是同步立即发的。
//   * `unsubscribe(topic):1317-1319` 只 remove，**不**发心跳。
//
// 为什么必须锁死：真机上「新订阅没推给 broker」是静默的 —— 只表现为
// QUERY_TOPIC_CONSUME_BY_WHO(300) 查不到本组、新 topic 分不到队列，客户端一声不响。
//
// 离线能测到哪一步：这里让消费者对着一个**连不上**的 name server 启动（路由拉不到 ⇒
// knownBrokerAddrs 为空 ⇒ 心跳一台都发不出去），因此只能证明
//   ① start() 之后 subscribe 不再抛「already started」；
//   ② 新订阅立刻进活订阅表（`subscribedTopics()`，也就是心跳与 rebalance 读的那张表）；
//   ③ unsubscribe 照旧只删表项。
// 报文层面「broker 真收到了带新订阅的心跳」由 examples/live_subscribe.cpp 在真机验证
// （broker 侧 topicGroupTable 的 300 号查询是唯一可信观测）。
#include <algorithm>
#include <chrono>
#include <cstdio>
#include <memory>
#include <string>
#include <thread>
#include <vector>

#include "rocketmq/client/consumer.h"
#include "rocketmq/client/exception.h"
#include "rocketmq/client/result.h"
#include "rocketmq/common/message.h"

using namespace rocketmq;

namespace {

int gChecks = 0;
int gFails = 0;

void check(bool ok, const std::string& name, const std::string& detail = std::string()) {
    ++gChecks;
    if (!ok) {
        ++gFails;
        std::printf("FAIL %s %s\n", name.c_str(), detail.c_str());
    }
}

// 一个永远连不上的地址：路由拉取当场 ECONNREFUSED（catch 住只记 debug 日志），
// knownBrokerAddrs() 因此为空 —— 消费者能正常 start()，但一个包都发不出去。
const char* kDeadNamesrv = "127.0.0.1:1";

class NoopListener : public MessageListenerConcurrently {
public:
    ConsumeConcurrentlyStatus consumeMessage(const std::vector<MessageExt>&,
                                            ConsumeConcurrentlyContext&) override {
        return ConsumeConcurrentlyStatus::CONSUME_SUCCESS;
    }
};

std::unique_ptr<DefaultMQPushConsumer> startedConsumer(const std::string& group) {
    auto c = std::make_unique<DefaultMQPushConsumer>(group);
    c->setNamesrvAddr(kDeadNamesrv);
    c->subscribe("BaseTopic", "*");
    c->setMessageListener(std::make_shared<NoopListener>());
    c->start();
    return c;
}

bool hasTopic(const std::vector<std::string>& topics, const std::string& topic) {
    return std::find(topics.begin(), topics.end(), topic) != topics.end();
}

// ① 后置 subscribe：Java 允许，且立即进活订阅表（旧实现在这里抛 already started）。
void testSubscribeAfterStartIsAllowed() {
    auto c = startedConsumer("GID_SubAfterStartUnit");
    check(c->isStarted(), "consumer is started");

    bool threw = false;
    std::string err;
    try {
        c->subscribe("LateTopic", "TagA||TagB");
    } catch (const std::exception& e) {
        threw = true;
        err = e.what();
    }
    check(!threw, "subscribe after start does not throw", threw ? err : "");
    check(hasTopic(c->subscribedTopics(), "LateTopic"),
          "late subscription lands in the live subscription table");

    // MessageSelector 重载同样允许（Java :1289-1303）。
    MessageSelector sql;
    sql.type = ExpressionType::SQL92;
    sql.expression = "a > 1";
    threw = false;
    try {
        c->subscribe("LateSqlTopic", sql);
    } catch (const std::exception& e) {
        threw = true;
        err = e.what();
    }
    check(!threw, "selector subscribe after start does not throw", threw ? err : "");
    check(hasTopic(c->subscribedTopics(), "LateSqlTopic"),
          "late selector subscription lands in the live subscription table");

    // 零 broker：心跳一台都发不出去，计数器不动（这条同时是「上面没偷偷发包」的对照）。
    check(c->heartbeatCount() == 0,
          "no broker known → heartbeat count stays 0",
          std::to_string(c->heartbeatCount()));

    c->shutdown();
    check(!c->isStarted(), "consumer stopped");
}

// ② unsubscribe 不发心跳（Java :1317-1319）：计数器不动，表项消失。
void testUnsubscribeOnlyDropsTheEntry() {
    auto c = startedConsumer("GID_SubUnsubUnit");
    c->subscribe("LateTopic", "*");
    const int64_t before = c->heartbeatCount();

    c->unsubscribe("LateTopic");

    check(c->heartbeatCount() == before, "unsubscribe does not push a heartbeat");
    check(!hasTopic(c->subscribedTopics(), "LateTopic"), "unsubscribed topic is gone");
    check(hasTopic(c->subscribedTopics(), "BaseTopic"), "other subscriptions untouched");
    c->shutdown();
}

// ③ 启动前订阅照旧（回归底线）：老用法不受影响。
void testSubscribeBeforeStartStillWorks() {
    DefaultMQPushConsumer c("GID_SubBeforeStartUnit");
    c.subscribe("T_A", "*");
    check(hasTopic(c.subscribedTopics(), "T_A"), "pre-start subscribe recorded");
    check(c.heartbeatCount() == 0, "pre-start subscribe sends nothing");
}

}  // namespace

int main() {
    testSubscribeBeforeStartStillWorks();
    testSubscribeAfterStartIsAllowed();
    testUnsubscribeOnlyDropsTheEntry();

    std::printf("%s: %d checks, %d failed\n", gFails == 0 ? "PASS" : "FAIL", gChecks, gFails);
    return gFails == 0 ? 0 : 1;
}
